packages/wrangler/src/experimental-assets.ts

```diff
@@ -1,8 +1,299 @@
+import assert from "node:assert";
 import { existsSync } from "node:fs";
-import path from "node:path";
-import { UserError } from "./errors";
+import { readdir, readFile, stat } from "node:fs/promises";
+import * as path from "node:path";
+import chalk from "chalk";
+import { getType } from "mime";
+import PQueue from "p-queue";
+import prettyBytes from "pretty-bytes";
+import { fetchResult } from "./cfetch";
+import { formatTime } from "./deploy/deploy";
+import { FatalError, UserError } from "./errors";
+import { logger, LOGGER_LEVELS } from "./logger";
+import {
+	BULK_UPLOAD_CONCURRENCY,
+	MAX_ASSET_COUNT,
+	MAX_ASSET_SIZE,
+	MAX_UPLOAD_ATTEMPTS,
+	MAX_UPLOAD_GATEWAY_ERRORS,
+} from "./pages/constants";
+import { hashFile } from "./pages/hash";
+import { APIError } from "./parse";
 import type { Config } from "./config";
 
+type AssetManifest = { [path: string]: { hash: string; size: number } };
+type InitializeAssetsResponse = {
+	// string of file hashes per bucket
+	buckets: string[][];
+	jwt: string;
+};
+
+export type UploadPayloadFile = {
+	name: string;
+	hash: string;
+	contents: string;
+	metadata: {
+		contentType: string;
+	};
+};
+
+type UploadResponse = {
+	jwt?: string;
+};
+
+export const syncAssets = async (
+	accountId: string | undefined,
+	scriptName: string,
+	assetDirectory: string | undefined,
+	dryRun: boolean | undefined
+): Promise<{
+	manifest: AssetManifest | undefined;
+	jwt: string | undefined;
+}> => {
+	if (assetDirectory === undefined) {
+		return { manifest: undefined, jwt: undefined };
+	}
+	if (dryRun) {
+		logger.log("(Note: doing a dry run, not uploading or deleting anything.)");
+		return { manifest: undefined, jwt: undefined };
+	}
+	assert(accountId, "Missing accountId");
+
+	const directory = path.resolve(assetDirectory);
```

Comment:
technically the `directory` path should already be resolved here. That's done in `getExperimentalAssetsBasePath` which should have been called somewhere in `deploy` I reckon. That was in fact the whole idea there, that `getExperimentalAssetsBasePath` gives us everything we need, so we don't have to worry about resolving paths and what not further down the stack. I wonder if we need to make that somewhat more clear 🤔 ?

Reply:
good point, just an oversight on my part :)